ipn/conffile/serveconf.go

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !ts_omit_serve
+
+package conffile
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"strings"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+	"tailscale.com/tailcfg"
+	"tailscale.com/types/opt"
+	"tailscale.com/util/mak"
+)
+
+// ServicesConfigFile is the config file format for services configuration.
+type ServicesConfigFile struct {
+	// 0.0.1, always present.
+	Version string `json:"version"`
+
+	Services map[tailcfg.ServiceName]ServiceDetailsFile `json:"services,omitzero"`
+}
+
+// ServiceDetailsFile is the config syntax for an individual Tailscale Service.
+type ServiceDetailsFile struct {
+	// 0.0.1, set if and only if this is not inside a [ServiceConfigFile].
+	Version string `json:"version,omitzero"`
+
+	Endpoints  map[*tailcfg.ProtoPortRange]*Target `json:"endpoints"`
```

Comment:
pointer to ProtoPortRange? Not a value?

Reply:
`UnmarshalText` is defined on a pointer receiver so unless I do this it doesn't serialize